workitem/workitem.go

```diff
@@ -53,3 +57,112 @@ func (wi WorkItem) GetLastModified() time.Time {
 	log.Debug(nil, map[string]interface{}{"wi_id": wi.ID}, "Last modified value: %v", lastModified)
 	return *lastModified
 }
+
+// ChangeSet derives a changeset between this workitem and a given workitem.
+func (wi WorkItem) ChangeSet(older convert.ChangeDetector) ([]convert.Change, error) {
+	if older == nil {
+		// this is changeset for a new ChangeDetector, report all observed attributes to
+		// the change set. This needs extension once we support more attributes.
+		changeSet := []convert.Change{
+			{
+				AttributeName: SystemState,
+				NewValue:      wi.Fields[SystemState],
+				OldValue:      nil,
+			},
+		}
+		if wi.Fields[SystemBoardcolumns] != nil && len(wi.Fields[SystemBoardcolumns].([]interface{})) != 0 {
+			changeSet = append(changeSet, convert.Change{
+				AttributeName: SystemBoardcolumns,
+				NewValue:      wi.Fields[SystemBoardcolumns],
+				OldValue:      nil,
+			})
+		}
+		return changeSet, nil
+	}
+	olderWorkItem, ok := older.(WorkItem)
+	if !ok {
+		return nil, errs.New("Other entity is not a WorkItem: " + reflect.TypeOf(older).String())
+	}
+	if wi.ID != olderWorkItem.ID {
+		return nil, errs.New("Other entity has not the same ID: " + olderWorkItem.ID.String())
+	}
+	changes := []convert.Change{}
+	// CAUTION: we're only supporting changes to the system.state and to the
+	// board position relationship for now. If we need to support more
```

Comment:
> If we need to support more attribute changes, this has to be added here.

As I've mentioned above, this should be done automatically and not by hand because we want to have actions on user-defined fields at some day and then we cannot modify the code for all fields. So far I cannot see anything that we couldn't do by taking the field type into account.

Reply:
I agree, but this is not needed here nor do we have a usecase to extend this and future usecases may be requiring a bigger scope than what we can project yet. Or in other words, the more we extend this beyond the MVP, the more the risk that we write code that need replacement later.

This here is doing exactly what we currently need, not more and not less. A full generic implementation would require considerably more work and may also be not matching the usecases later. So I don't see a reasonable reason to extend that (yet).

Reply:
I partially agree. We might need to touch more than just the fields of a work item but still, this is the most common area for work items. The reason I'm worried so much is that we can also not foresee how much pain it causes when we have hard-coded field handling in some places. It cause a lot of problems for me in the past. But I'm fine with it for now. All I'm saying is that we shouldn't start adding more fields here manually but instead rework this at the moment, the next field is due. Okay?